contracts/PercentageTransferManager/src/lib.rs

```diff
@@ -0,0 +1,593 @@
+                    (2000u128 * multiplier).into(),
+                ),
+                RestrictionResult::Invalid
+            );
+
+            // Should fail when the balance of will be more the restriction
+            assert_eq!(
+                percentage_transfer_manager.verify_transfer(
+                    Some(from),
+                    Some(to),
+                    (301u128 * multiplier).into(),
+                    (2000u128 * multiplier).into(),
+                    (100u128 * multiplier).into(),
+                    (2000u128 * multiplier).into()
+                ),
+                RestrictionResult::Invalid
+            );
+
+            // Should fail when primary issuance is on because from is not None
+            percentage_transfer_manager.change_primary_issuance(true);
+            // check for the primary issuance value
+            assert_eq!(
+                percentage_transfer_manager.is_primary_issuance_allowed(),
+                true
+            );
+            assert_eq!(
+                percentage_transfer_manager.verify_transfer(
+                    Some(from),
+                    Some(to),
+                    (700u128 * multiplier).into(),
+                    (2000u128 * multiplier).into(),
+                    0u128.into(),
+                    (2000u128 * multiplier).into()
+                ),
+                RestrictionResult::Invalid
+            );
+
+            // Should pass when primary issuance is on & from is None
+            assert_eq!(
+                percentage_transfer_manager.verify_transfer(
+                    None,
+                    Some(to),
+                    (700u128 * multiplier).into(),
+                    (2000u128 * multiplier).into(),
+                    0u128.into(),
+                    (2000u128 * multiplier).into()
+                ),
+                RestrictionResult::Valid
+            );
+
+            // Should pass when the Identity in the exemption list
+            percentage_transfer_manager.change_primary_issuance(false);
+            assert_eq!(
+                percentage_transfer_manager.is_primary_issuance_allowed(),
+                false
+            );
+            percentage_transfer_manager.modify_exemption_list(to, true);
+            assert_eq!(percentage_transfer_manager.is_exempted_or_not(to), true);
+
+            assert_eq!(
+                percentage_transfer_manager.verify_transfer(
+                    None,
+                    Some(to),
+                    (700u128 * multiplier).into(),
+                    (2000u128 * multiplier).into(),
+                    0u128.into(),
+                    (2000u128 * multiplier).into()
+                ),
+                RestrictionResult::Valid
+            );
+        }
+
+        #[test]
+        fn test_verify_transfer_with_decimal_percentage() {
+            let percentage_transfer_manager = PercentageTransferManagerStorage::new(278940, false); // it is 27.894% of the totalSupply
```

Comment:
Could be something like `FixedPrecision::from_bits(268940)`, like in https://docs.rs/fixed/0.5.3/fixed/struct.FixedU128.html
I'm not sure if the type could just be `FixedU128::<U6>` or its wrapper because of the requirement to make the `u128` public.

Reply:
I don't think FixedU128 implementation is provided by the ink!. We have to create a custom type for it. Let's keep it simple for now and we can add these types in the future if needed. 
It is better to have some JIRA story that will cover this research.

Reply:
Agreed, `FixedU128` looks like an overkill.